extensions/blocks/podcast-player/podcast-player.js

```diff
@@ -0,0 +1,372 @@
+		if ( ! trackData ) {
+			return;
+		}
+		this.audio.src = trackData.src;
+		this.currentTrack = track;
+		this.play();
+		this.setTrackState( track, true );
+	}
+
+	/**
+	 * Visually mark track active or inactive.
+	 * @private
+	 * @param {number} track The track number
+	 * @param {boolean} isActive True if track is active
+	 * @param {boolean} isError True if track couldn't be played
+	 */
+	setTrackState( track, isActive, isError = false ) {
+		// Mark track element as active/inactive.
+		const el = this.trackElements[ track ];
+
+		// Active state classname.
+		if ( isActive ) {
+			el.classList.add( 'is-active' );
+		} else {
+			el.classList.remove( 'is-active' );
+		}
+
+		// Maintain pressed state.
+		el.querySelector( '[aria-pressed]' ).setAttribute( 'aria-pressed', JSON.stringify( isActive ) );
+
+		// Add or remove error.
+		const errorEl = el.querySelector( '.podcast-player__episode-error' );
+		if ( isError && ! errorEl ) {
+			const trackData = this.getTrack( track );
+			el.appendChild( this.renderErrorWithLink( trackData && trackData.link ) );
+		} else if ( ! isError && errorEl ) {
+			errorEl.remove();
+		}
+
+		const iconEl = el.querySelector( '.podcast-player__episode-status-icon' );
+		if ( iconEl ) {
+			if ( isActive ) {
+				if ( isError ) {
+					iconEl.innerHTML = errorIcon;
+				} else {
+					iconEl.innerHTML = soundIcon;
+				}
+			} else {
+				iconEl.innerHTML = '';
+			}
+		}
+	}
+
+	/**
+	 * Construct error markup.
+	 * @private
+	 * @param {string} link Optional URL to use as a fallback link
+	 * @returns {Element} Error DOM Element
+	 */
+	renderErrorWithLink( link ) {
+		const errorEl = document.createElement( 'div' );
+		errorEl.classList.add( 'podcast-player__episode-error' );
+		errorEl.appendChild( new Text( __( 'Episode unavailable', 'jetpack' ) ) );
+
+		if ( link ) {
+			const linkEl = document.createElement( 'a' );
+			linkEl.rel = 'noopener noreferrer nofollow';
+			linkEl.target = '_blank';
+			linkEl.href = link;
+			linkEl.innerText = __( 'Open in a new tab', 'jetpack' );
+
+			const spanEl = document.createElement( 'span' );
+			spanEl.appendChild( new Text( ' (' ) );
+			spanEl.appendChild( linkEl );
+			spanEl.appendChild( new Text( ')' ) );
+			errorEl.appendChild( spanEl );
+		}
+
+		return errorEl;
+	}
+
+	/**
+	 * Walk the DOM tree up and check whether it is inside
+	 * an element representing a track.
+	 * @private
+	 * @param {Element} el Element to check
+	 * @returns {mumber} The track number or -1
+	 */
+	getTrackFromElement( el ) {
+		const track = el.closest( '.podcast-player__episode' );
+		return track ? this.trackElements.indexOf( track ) : -1;
+	}
+
+	/**
+	 * Handle pressing a key.
+	 * @private
+	 * @param {KeyboardEvent} e The event
+	 */
+	handleKeyDown( e ) {
+		// Only handle the Space key.
+		if ( event.key !== ' ' ) {
+			return;
+		}
+
+		// Check if we are inside a track element and which one.
+		const track = this.getTrackFromElement( e.target );
```

Comment:
Since this ends up calling `el.closest( '.podcast-player__episode' )`, will it end up walking up the entire DOM tree? If so, is there a way we can return earlier so spacebar presses outside of the podcast player don't trigger an entire DOM traverse? Or is it OK since we're only registering this keydown listener inside the `block` now instead of the previous `document` listener?

Reply:
It ends up walking, you are right. However, this event should only trigger when you are focused somewhere inside the block (because it's bound to its element) so I think it's guaranteed to always find the element and never break out of it.